File: TwoSuM_optimal.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> sortedNums = nums;
        sort(sortedNums.begin(), sortedNums.end());

        int left = 0;
        int right = sortedNums.size() - 1;

        while (left <= right) {
            int sum = sortedNums[left] + sortedNums[right];

            if (sum == target) {
                return {left, right};
            }

            if (sum < target) {
                left++;
            } else {
                right--;
            }
        }

        return {};
    }
};

void printVector(const vector<int>& values) {
    cout << "[";
    for (int i = 0; i < values.size(); i++) {
        cout << values[i];
        if (i != values.size() - 1) {
            cout << ", ";
        }
    }
    cout << "]\n";
}

int main() {
    Solution sol;

    vector<int> nums1 = {2, 7, 11, 15};
    vector<int> nums2 = {3, 3};
    vector<int> nums3 = {3, 2, 4};

    printVector(sol.twoSum(nums1, 9));
    printVector(sol.twoSum(nums2, 6));
    printVector(sol.twoSum(nums3, 7));

    return 0;
}

File: TwoSuM_optimal.py

# This can be only used for like sayin Yes or No becosue we are not storing any index properly
from typing import List 
class Solution:
    def two_sum(self,nums: List[int], target: int) -> List[int]:
       sorted_nums = sorted(nums)
       l =0
       r =len(nums)-1
       while l <= r:
            if sorted_nums[l] + sorted_nums[r] == target:
                return [l,r]
            elif sorted_nums[l] + sorted_nums[r] < target:
                l+=1
            else:
                r-=1

           

               


sol = Solution()
print(sol.two_sum([2,7,11,15],9))
print(sol.two_sum([3,3],6))
print(sol.two_sum([3,2,4],7))